src/Microsoft.VisualStudio.Validation/Requires.cs

```diff
@@ -28,7 +30,7 @@ public static class Requires
         /// <exception cref="ArgumentNullException">Thrown if <paramref name="value"/> is <c>null</c>.</exception>
         [DebuggerStepThrough]
         [TargetedPatchingOptOut("Performance critical to inline across NGen image boundaries")]
-        public static T NotNull<T>([ValidatedNotNull, NotNull]T value, string parameterName)
+        public static T NotNull<T>([ValidatedNotNull, NotNull]T? value, string? parameterName)
```

Comment:
I wonder if this is really the right approach. Maybe I should state that these methods *don't* permit a null ref type. I want to accomplish two objectives:

1. If C# code calls this method with a `T?` type, the compiler will know if the method returns that it's definitely just `T` (non-null) and thus subsequent code doesn't produce warnings when it is dereferenced.
1. If the C# code calls this method, passing in a parameter from the caller's method, and if it is typed as `T?`, I suppose they should get a warning. We actually want folks to change their `T?` to `T` if they call `Requires.NotNull` with that parameter, so that they're documenting that null arguments are actually not allowed.

Reply:
Ya, I think I should remove the `?` nullable annotation for all methods that will throw if it is null, even if that is their express purpose.

Reply:
It's an interesting decision. At first brush I leaned the opposite way (permitting nullable input), but I _think_ I agree that it's more consistent to force `!` in all places where you're overriding the type system's analysis.

The difference here is that before you'd `!` the `Requires.NotNull` argument, then have to `!` the first usage after that:

```c#
void M(string? s)
{
    Requires.NotNull(s!);
    s!.ToString(); // null forgiveness here too
}
```

Now you'll just have to `!` the argument, which avoids repeating yourself:

```c#
void M(string? s)
{
    Requires.NotNull(s!);
    s.ToString();
}
```

---

Would you feel differently about the `Assumes.NotNull` case?

Reply:
If I change this back to `[NotNull] T value`:
1. That encourages callers that have nullable enabled to remove the `?` from their parameter type, which helps propagate the non-nullability. Good thing.
1. For callers that want to keep `T?` as their parameter type, they can use `Requires.NotNull(v!, nameof(v));` to override the warning.
1. After this call they can immediately dereference `v`  without a warning because `[NotNull]` on our parameter tells the compiler we won't return if it is null. No extra `!` needed.
1. It makes no difference to callers that don't have nullable enabled.

And ya, I was divided on this earlier too, since these methods are explicitly designed to correctly process both null and non-null inputs. However it's pretty clear now that we analyze it that the desirable behavior comes from removal of the `?`.
And yes, I'm glad you raised the question regarding whether `Assumes` should be treated differently. I think you're right. There's no propagation of non-nullable parameter types that we want to push outward for the `Assumes` case, presumably, so accepting `?` just makes it easier to call it.